src/vcpkg/base/downloads.cpp

```diff
@@ -28,7 +28,45 @@ namespace vcpkg
         return input;
     }
 
+    bool operator==(const ProxyCredentials& lhs, const ProxyCredentials& rhs)
+    {
+        return lhs.username == rhs.username && lhs.password == rhs.password;
+    }
+
+    bool operator==(const ProxyUrlParts& lhs, const ProxyUrlParts& rhs)
+    {
+        return lhs.host == rhs.host && lhs.credentials == rhs.credentials;
+    }
+
+    ProxyUrlParts parse_proxy_url(const std::wstring& url)
+    {
+        constexpr auto npos = std::wstring::npos;
+        auto scheme_separator_pos = url.find(L"://");
+        std::wstring scheme;
+        std::wstring_view remaining_parts = url;
+        if (scheme_separator_pos != npos)
+        {
+            scheme = remaining_parts.substr(0, scheme_separator_pos + 3);
+            remaining_parts = remaining_parts.substr(scheme_separator_pos + 3);
+        }
+        Optional<ProxyCredentials> credentials;
+        auto credentials_separator_pos = remaining_parts.find(L"@");
+        auto password_separator_pos = remaining_parts.find(L":");
```

Comment:
This seems incorrect as the password separator, if present, must be inside the userinfo. Perhaps the STL algorithm form would be clearer since it makes chopping ranges into subranges like this easy:

```c++
auto user_info_end = std::find(remaining_parts.begin(), remaining_parts.end(), '@');
if (user_info_end != remaining_parts.end()) {
    auto password_separator = std::find(remaining_parts.begin(), user_info_end, ':');
    if (password_separator != user_info_end) {
        auto password_start = password_separator + 1;
        credentials = ProxyCredentials{std::wstring(remaining_parts.begin(), password_separator - remaining_parts.begin()),
        std::wstring(password_start, user_info_end - password_start)};
    }
}
```

Also, you want the character finds (`L'@'`) not string finds (`L"@"`).